src/monitor/circuit_graph.rs

```diff
@@ -0,0 +1,379 @@
+use std::{
+    borrow::Cow,
+    collections::{hash_map::Entry, HashMap},
+    slice,
+};
+
+use super::visual_graph::{
+    ClusterNode, Edge as VisEdge, Graph as VisGraph, Node as VisNode, SimpleNode,
+};
+use crate::circuit::{trace::EdgeKind, GlobalNodeId, NodeId};
+
+/// A region is a named grouping of operators in a circuit.
+///
+/// Regions can be nested inside other regions, forming a tree.
+/// A circuit is created with a single root region.
+pub(super) struct Region {
+    id: RegionId,
+    pub(super) nodes: Vec<NodeId>,
+    name: Cow<'static, str>,
+    children: Vec<Region>,
+}
+
+/// Region id is a path from the root of the region tree.
+#[repr(transparent)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct RegionId(Vec<usize>);
+
+impl RegionId {
+    /// Root region.
+    pub(super) fn root() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Pop the innermost child, transforming region id into its parent region
+    /// id.
+    pub(super) fn pop(&mut self) {
+        self.0.pop();
+    }
+
+    pub(super) fn child(&self, child_id: usize) -> Self {
+        let mut path = Vec::with_capacity(self.0.len() + 1);
+        path.extend_from_slice(&self.0);
+        path.push(child_id);
+        Self(path)
+    }
+}
+
+impl Region {
+    pub(super) fn new(id: RegionId, name: Cow<'static, str>) -> Self {
+        Self {
+            id,
+            nodes: Vec::new(),
+            name,
+            children: Vec::new(),
+        }
+    }
+
+    /// Generate unique name for a region to use as a node label in a visual
+    /// graph.
+    fn region_identifier(node_id: &GlobalNodeId, region_id: &RegionId) -> String {
+        let mut region_ident = format!(
+            "{}{}",
+            Node::node_identifier(node_id),
+            if region_id.0.is_empty() { "" } else { "_r" }
+        );
+
+        for i in 0..region_id.0.len() {
+            region_ident.push_str(&region_id.0[i].to_string());
+            if i < region_id.0.len() - 1 {
+                region_ident.push('_');
+            }
+        }
+        region_ident
+    }
+
+    /// Output region as a cluster in a visual graph.
+    fn visualize<F>(&self, scope: &Node, annotate: &F) -> ClusterNode
+    where
+        F: Fn(&GlobalNodeId) -> String,
```

Comment:
Don't take a reference to a closure and take an `FnMut`
```suggestion
    fn visualize<F>(&self, scope: &Node, mut annotate: F) -> ClusterNode
    where
        F: FnMut(&GlobalNodeId) -> String,
```